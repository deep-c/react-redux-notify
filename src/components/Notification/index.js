import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styleMap from './Notification.scss';


export class Notification extends React.PureComponent {

  static propTypes = {
    id: PropTypes.number.isRequired,
    type: PropTypes.string.isRequired,
    canDismiss: PropTypes.bool.isRequired,
    duration: PropTypes.number.isRequired,
    icon: PropTypes.node,
    customStyles: PropTypes.object,
    customComponent: PropTypes.element,
    acceptBtn: PropTypes.shape({
      handler: PropTypes.func.isRequired,
      icon: PropTypes.node,
      title: PropTypes.node,
    }),
    denyBtn: PropTypes.shape({
      handler: PropTypes.func.isRequired,
      icon: PropTypes.node,
      title: PropTypes.node,
    }),
    isFirst: PropTypes.bool.isRequired,
    handleDismiss: PropTypes.func.isRequired,
    handleDismissAll: PropTypes.func.isRequired,
    styles: PropTypes.object.isRequired,
  }

  static defaultProps = {
    styles: styleMap,
    canDismiss: true,
    duration: 0,
  }

  componentDidMount() {
    const { handleDismiss, id, duration } = this.props;
    if (duration !== 0) {
      setTimeout(() => { handleDismiss(id); }, duration);
    }
  }

  render() {
    const { handleDismiss, handleDismissAll, isFirst, message, type, canDismiss,
      acceptBtn, denyBtn, icon, customStyles, id } = this.props;
    let { styles } = this.props;
    styles = Object.assign({}, styles, customStyles);
    const cx = classNames.bind(styles);
    const containerTypeClass = cx({
      'has-close': !isFirst && canDismiss,
      'no-close': !isFirst && !canDismiss,
      'has-close-all': isFirst && canDismiss,
      'has-close-all--noDismiss': isFirst && !canDismiss,
      [`notification--${type.toLowerCase()}`]: true,
    });

    return (
      <div key={id} className={containerTypeClass}>
        {(icon) ? <span className={styles.icon}>{icon}</span> : false}
        <div className={styles.content}>
          <div className={styles['item--message']}>{message}</div>
          { (!canDismiss && (acceptBtn || denyBtn)) ?
              <div className={styles['item--btnBar']}>
                {(acceptBtn) ?
                  <div className={styles.actionBtn}
                    onClick={(e) => {
                      acceptBtn.handler(e, this.props);
                    }}>
                    {(acceptBtn.icon && typeof acceptBtn.icon === 'string') ?
                      <i className={acceptBtn.icon} />
                      :
                      acceptBtn.icon
                    }
                    {acceptBtn.title}
                  </div>
                  : false
                }
                {(denyBtn) ?
                  <div className={styles.actionBtn}
                    onClick={(e) => {
                      denyBtn.handler(e, this.props);
                    }}>
                    {(denyBtn.icon && typeof denyBtn.icon === 'string') ?
                      <i className={denyBtn.icon} />
                      :
                      denyBtn.icon
                    }
                    {denyBtn.title}
                  </div>
                  :
                  false
                }
              </div>
              :
              false
          }
        </div>
        {(canDismiss) ?
          <div className={styles.close} onClick={() => handleDismiss(id)}></div>
          :
          false
        }
        {(isFirst && canDismiss) ?
          <div className={styles['close-all']}
            onClick={() => handleDismissAll()}>Close All</div>
          : false
        }
      </div>
    );
  }
}

export default Notification;
