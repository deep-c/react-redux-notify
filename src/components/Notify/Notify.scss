$width: 300px;
$max-height: 150px;

@mixin container {
  z-index: 9998;
  position: fixed;
  overflow-y: auto;
  max-height: 100vh;
  &::-webkit-scrollbar {
    display: none;
  }     
}

@keyframes notification-show {
    0% { opacity: 0; transform: perspective($width) translate(0, -30px) rotateX(90deg); }
  100% { opacity: 1; transform: perspective($width) translate(0, 0) rotateX( 0deg); }
}

@keyframes notification-hide {
    0% { opacity: 1; transform: scale( 1); }
  100% { opacity: 0; transform: scale(.8); }
}

@keyframes notification-shrink {
    0% {  opacity: 0; max-height: $max-height; transform: scale(.8); }
  100% {  opacity: 0; max-height: 0;           transform: scale(.8); }
}

.wrapper {
  display: flex;  
  padding: 10px;
  font-size: 1.2em;
  min-height: 0;
  flex-direction: inherit       
}

.containerTopRight {
  @include container;
  top: 0;
  right: 0;
  flex-direction: column;
}

.containerBottomRight {
  @include container;
  right: 0;
  bottom: 0;
  flex-direction: column-reverse;
}

.containerBottomLeft {
  @include container;
  left: 0;
  bottom: 0;
  flex-direction: column-reverse;
}

.containerTopLeft {
  @include container;
  top: 0;
  left: 0;
  flex-direction: column;
}

.enter {
  animation: notification-show .16s cubic-bezier(0.175, 0.885, 0.32, 1.27499);
}

.leave {
  animation: notification-hide   .24s      cubic-bezier(0.33859, -0.42, 1, -0.22),
             notification-shrink .24s .24s cubic-bezier(0.5, 0, 0, 1);
}

