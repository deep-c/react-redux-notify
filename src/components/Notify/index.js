import React from 'react';
import PropTypes from 'prop-types';
import { createPortal } from 'react-dom';
import { connect } from 'react-redux';
import ReactCSSTransitionGroup from 'react-transition-group/CSSTransitionGroup';
import {
  removeNotification,
  removeAllNotifications,
  NOTIFICATIONS_POS_TOP_RIGHT,
} from 'modules/Notifications';
import { default as Notification } from 'components/Notification';
import { canUseDOM } from 'utils';
import styles from './Notify.scss';

export class Notify extends React.PureComponent {
  static propTypes = {
    notifications: PropTypes.array.isRequired,
    remove: PropTypes.func.isRequired,
    removeAll: PropTypes.func.isRequired,
    customStyles: PropTypes.object,
    notificationComponent: PropTypes.func,
    transitionDurations: PropTypes.shape({
      enter: PropTypes.number,
      leave: PropTypes.number,
    }),
    position: PropTypes.string,
    forceClose: PropTypes.bool,
    localization: PropTypes.shape({
      closeAllBtnText: PropTypes.string,
      acceptBtnText: PropTypes.string,
      denyBtnText: PropTypes.string,
    }),
    node: PropTypes.any,
  };

  static defaultProps = {
    notificationComponent: Notification,
    transitionDurations: {
      enter: 160,
      leave: 400,
    },
    position: NOTIFICATIONS_POS_TOP_RIGHT,
    customStyles: {},
    forceClose: false,
    localization: {
      closeAllBtnText: 'Close All',
      acceptBtnText: 'Accept',
      denyBtnText: 'Deny',
    },
  };

  constructor(props) {
    super(props);
    this.handleDismiss = this._handleDismiss.bind(this);
    this.handleDismissAll = this._handleDismissAll.bind(this);
  }

  componentWillUnmount() {
    if (this.defaultNode) {
      // eslint-disable-next-line no-undef
      document.body.removeChild(this.defaultNode);
    }
    this.defaultNode = null;
  }

  _handleDismiss(id) {
    const { remove } = this.props;
    remove(id);
  }

  _handleDismissAll(force) {
    const { removeAll, forceClose } = this.props;
    removeAll(force || forceClose);
  }

  _getStyle(name) {
    return this.props.customStyles[name] || styles[name];
  }

  _render() {
    const {
      notifications,
      notificationComponent,
      transitionDurations,
      position,
      localization,
    } = this.props;
    const notificationsContainerClass = this._getStyle(`container${position}`);

    return (
      <div className={notificationsContainerClass}>
        <ReactCSSTransitionGroup
          component="div"
          className={this._getStyle('wrapper')}
          transitionName={{
            enter: this._getStyle('enter'),
            leave: this._getStyle('leave'),
          }}
          transitionEnterTimeout={transitionDurations.enter}
          transitionLeaveTimeout={transitionDurations.leave}
        >
          {notifications.map((notification, i) => {
            const NewNotification =
              notification.customComponent || notificationComponent;
            return (
              <NewNotification
                key={notification.id}
                localization={localization}
                {...notification}
                isFirst={i === 0 && notifications.length > 1}
                handleDismiss={this.handleDismiss}
                handleDismissAll={this.handleDismissAll}
              />
            );
          })}
        </ReactCSSTransitionGroup>
      </div>
    );
  }

  render() {
    const { node } = this.props;
    /* istanbul ignore if  */
    if (!canUseDOM) {
      return null;
    }
    /* istanbul ignore if  */
    if (!node && !this.defaultNode) {
      /* eslint-disable no-undef */
      this.defaultNode = document.createElement('div');
      document.body.appendChild(this.defaultNode);
    }

    return createPortal(this._render(), node || this.defaultNode);
  }
}

/* istanbul ignore next */
const mapStateToProps = state => ({
  notifications: state.notifications,
});

/* istanbul ignore next */
const mapDispatchToProps = dispatch => ({
  remove: (id) => {
    dispatch(removeNotification(id));
  },
  removeAll: (force) => {
    dispatch(removeAllNotifications(force));
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(Notify);
