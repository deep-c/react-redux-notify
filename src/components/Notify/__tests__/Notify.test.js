import React from 'react';
import { mount } from 'enzyme';
import { Notify } from 'components/Notify';
import {
  NOTIFICATION_TYPE_SUCCESS,
  NOTIFICATION_TYPE_INFO,
  NOTIFICATION_TYPE_ERROR,
} from 'modules/Notifications';

describe('Notify', () => {
  const handleRemove = jest.fn();
  const handleRemoveAll = jest.fn();
  const notification1 = {
    id: 1,
    message: 'Notification 1!',
    type: NOTIFICATION_TYPE_SUCCESS,
    duration: 0,
    canDismiss: true,
    icon: <i className="fa fa-check" />,
  };
  const notification2 = {
    id: 2,
    message: 'Notification 2!',
    type: NOTIFICATION_TYPE_INFO,
    duration: 0,
    canDismiss: true,
    icon: <i className="fa fa-check" />,
  };
  const notification3 = {
    id: 3,
    message: 'Notification 3!',
    type: NOTIFICATION_TYPE_ERROR,
    duration: 0,
    canDismiss: true,
    icon: <i className="fa fa-check" />,
  };
  const props = {
    notifications: [],
    remove: handleRemove,
    removeAll: handleRemoveAll,
  };

  it('renders with required props', () => {
    const component = mount(<Notify {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('renders notifications', () => {
    const tProps = {
      ...props,
      notifications: [notification1, notification2],
    };
    const component = mount(<Notify {...tProps} />);
    // TODO: Fix this, not sure whats going on here, but children's length
    // doesn't seem to be correct, even though the output in the
    // snapshot is correct.
    // expect(component.find(".wrapper").children()).toHaveLength(2);
    expect(component).toMatchSnapshot();
  });

  it('renders custom notifications', () => {
    const MyCustomNotificationComponent = ({
      message,
      canDismiss,
      id,
      handleDismiss,
    }) => {
      let styles = {
        margin: '5px 0',
        padding: '2px 5px',
        border: '1px solid #333',
        float: 'right',
        clear: 'right',
        width: '330px',
        boxSizing: 'border-box',
      };
      if (canDismiss) {
        styles = Object.assign({}, styles, { cursor: 'pointer' });
      }
      return (
        <div
          onClick={() => {
            if (canDismiss) {
              handleDismiss(id);
            }
          }}
          style={styles}
        >
          {message}
        </div>
      );
    };
    const tProps = {
      ...props,
      notificationComponent: MyCustomNotificationComponent,
      notifications: [notification1],
    };
    const component = mount(<Notify {...tProps} />);
    expect(component).toMatchSnapshot();
    expect(component.contains(MyCustomNotificationComponent)).toEqual(true);
  });

  it('renders with the correct className when customStyles is used', () => {
    const tProps = {
      ...props,
      customStyles: {
        containerCustomPosition: 'CustomPosition',
      },
      position: 'CustomPosition',
    };
    const component = mount(<Notify {...tProps} />);
    expect(component.find('.CustomPosition')).toHaveLength(1);
    expect(component).toMatchSnapshot();
  });

  it('passes the correct props to each notification', () => {
    const tProps = {
      ...props,
      notifications: [notification1, notification2, notification3],
    };
    const component = mount(<Notify {...tProps} />);
    expect(component.find({ ...notification1, isFirst: true })).toHaveLength(1);
    expect(component.find({ ...notification2, isFirst: false })).toHaveLength(
      1
    );
    expect(component.find({ ...notification3, isFirst: false })).toHaveLength(
      1
    );
  });

  it('calls removeAll with the correct arguments', () => {
    const tProps = {
      ...props,
      notifications: [notification1, notification2, notification3],
    };
    const component = mount(<Notify {...tProps} />);
    component.find('.close-all').simulate('click');
    expect(handleRemoveAll).toHaveBeenCalled();
    component.setProps({ forceClose: true });
    component.find('.close-all').simulate('click');
    expect(handleRemoveAll).lastCalledWith(true);
  });

  it('calls remove with the correct arguments', () => {
    const tProps = { ...props, notifications: [notification1] };
    const component = mount(<Notify {...tProps} />);
    component
      .find('.close')
      .first()
      .simulate('click');
    expect(handleRemove).toHaveBeenCalled();
  });

  it('unmounts without error', () => {
    const component = mount(<Notify {...props} />);
    const instance = component.instance();
    expect(instance.defaultNode).not.toEqual(null);
    component.unmount();
    expect(instance.defaultNode).toEqual(null);
    expect(component).toMatchSnapshot();
  });
});
