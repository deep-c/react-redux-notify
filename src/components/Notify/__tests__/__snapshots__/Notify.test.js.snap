// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notify renders custom notifications 1`] = `
<div
  className="containerTopRight"
>
  <div
    className="wrapper"
  >
    <div
      onClick={[Function]}
      style={
        Object {
          "border": "1px solid #333",
          "boxSizing": "border-box",
          "clear": "right",
          "cursor": "pointer",
          "float": "right",
          "margin": "5px 0",
          "padding": "2px 5px",
          "width": "330px",
        }
      }
    >
      Notification 1!
    </div>
  </div>
</div>
`;

exports[`Notify renders notifications 1`] = `
<div
  className="containerTopRight"
>
  <div
    className="wrapper"
  >
    <div
      className="has-close-all notification--success"
    >
      <span
        className="icon"
      >
        <i
          className="fa fa-check"
        />
      </span>
      <div
        className="content"
      >
        <div
          className="item__message"
        >
          Notification 1!
        </div>
      </div>
      <div
        className="close"
        onClick={[Function]}
      />
      <div
        className="close-all"
        onClick={[Function]}
      >
        Close All
      </div>
    </div>
    <div
      className="has-close notification--info"
    >
      <span
        className="icon"
      >
        <i
          className="fa fa-check"
        />
      </span>
      <div
        className="content"
      >
        <div
          className="item__message"
        >
          Notification 2!
        </div>
      </div>
      <div
        className="close"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`Notify renders with required props 1`] = `
<div
  className="containerTopRight"
>
  <div
    className="wrapper"
  />
</div>
`;

exports[`Notify renders with the correct className when customStyles is used 1`] = `
<div
  className="CustomPosition"
>
  <div
    className="wrapper"
  />
</div>
`;

exports[`Notify unmounts without error 1`] = `""`;
